### ISSUE TITLE
Instability for UTM projection

### ISSUE THREAD

**Issue by perliedman** [proj4js/proj4js#187]
When repeatedly projecting and unprojecting a coordinate in UTM projection, proj4 seems to be surprisingly unstable on northern latitudes, about an order of magnitude more unstable than `cs2cs`.

I've run the following test program:

``` javascript
var proj4 = require('proj4'),
    proj = proj4(proj4.WGS84, process.argv[2]),
    ll = [31, 70],
    p0 = proj.forward(ll);

for (var i = 0; i < 10; i++) {
    p = proj.forward(ll);
    ll = proj.inverse(p);
}

var dx = p[0] - p0[0],
    dy = p[1] - p0[1];

console.log(Math.sqrt(dx*dx + dy*dy));
```

Running this with EPSG:32633 (`+proj=utm +zone=33 +datum=WGS84 +units=m +no_defs`), results in a difference of 364 meters.

Doing the same thing with cs2cs (https://gist.github.com/perliedman/44ea458607fba4f121314628c9cee978) gives a difference of 23 meters.

Is this difference expected? It causes trouble for some Leaflet users (https://github.com/Leaflet/Leaflet/issues/4480), since Leaflet internally will sometimes need to project and unproject the same coordinate.

**Comment by mbloch** [comment]
The greater stability of cs2cs is likely because it uses a more numerically stable implementation of transverse mercator for UTM than proj4js ("etmerc" instead of "tmerc"). Perhaps etmerc should be ported to proj4js.

**Comment by calvinmetcalf** [comment]
etmerc instead of tmerc would be very welcome, if anyone wants to open a pull

**Comment by theashyster** [comment]
@perliedman I am now looking more into this and will try to port etmerc to proj4js.
Now I am trying to get as much test data as possible and used your script.

> Doing the same thing with cs2cs (https://gist.github.com/perliedman/44ea458607fba4f121314628c9cee978) gives a difference of 23 meters.

I tried it an it gives me a difference of 0 meters, could this be correct? Which version of proj.4 did you use for your test back on 22 Apr 2016? Can you do a retest of this using the master code of proj.4 lib?

@mbloch I also tried using your https://github.com/mbloch/mapshaper-proj lib and using the mcs2cs tool and for me it looks like the results are more the same as using the cs2cs from proj.4 lib. I see that you have etmerc already ported to js in your lib, so I should probably be able to port it to proj4js from your lib, right? Any hints would help a ton.

**Comment by perliedman** [comment]
@theashyster Wow, great! I started looking at implementing etmerc, and it seemed doable, but other things got in the way, so this would be awesome.

Re: proj.4 version:

```shell
per@linfro:~/Documents/tmp$ cs2cs -v
Using from definition: 
Rel. 4.9.2, 08 September 2015
```

I'm running a bit low on time, so compiling latest Proj.4 is not something I'm keen on doing ATM. The difference is weird, maybe you could add some debug output so you can verify that it's actually converting values back and forth, and not just running them through an identity transform or something else?

**Comment by theashyster** [comment]
@perliedman :) Heh, yeah, seems doable, but will see if I can actually do it. I don't have that much of knowledge working with proj4js itself.

Thanks, hmm, seems you used version 4.9.2 which actually did not use etmerc yet, only starting from version 4.9.3 utm actually switched to using etmerc. So could it be that your script was actually still using utm with tmerc and cs2cs using tmerc returned a difference of 23 meters?

I see, understandable. Ok, will try that, thanks.

**Comment by perliedman** [comment]
Huh, gets even weirder. If my cs2cs doesn't run etmerc, why does it have better precision than the JS version?

Anyway, with some luck, etmerc could make this work with perfect precision.

**Comment by theashyster** [comment]
Definitely. Heh, don't know. I think, I will try this with cs2cs 4.9.2 and see what happens.
Yeah, hopefully.

Added some console.log to the loop in the script for cs2cs and the results returned are like this:
```
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
1104629.4356454364  7845845.0775721390 0.0000000000
31dE  70dN 0.000
0
```

**Comment by theashyster** [comment]
Okay, run some tests using @perliedman script for cs2cs and some logging added to it.

Using:
```
osboxes@osboxes:~/proj.4/samples$ cs2cs -v
Using from definition: 
Rel. 4.9.2, 08 September 2015
```

Returns:
```
1104629.43 7845845.08 0.00
30d59'59.765"E 70d0'0.003"N 0.000
1104626.99 7845844.52 0.00
30d59'59.53"E 70d0'0.006"N 0.000
1104624.55 7845843.95 0.00
30d59'59.295"E 70d0'0.009"N 0.000
1104622.11 7845843.39 0.00
30d59'59.06"E 70d0'0.012"N 0.000
1104619.66 7845842.83 0.00
30d59'58.824"E 70d0'0.015"N 0.000
1104617.21 7845842.27 0.00
30d59'58.588"E 70d0'0.018"N 0.000
1104614.76 7845841.70 0.00
30d59'58.352"E 70d0'0.021"N 0.000
1104612.31 7845841.14 0.00
30d59'58.117"E 70d0'0.024"N 0.000
1104609.87 7845840.58 0.00
30d59'57.882"E 70d0'0.027"N 0.000
1104607.43 7845840.02 0.00
30d59'57.647"E 70d0'0.03"N 0.000
22.574401431827113
```

Using:
```
osboxes@osboxes:~/proj.4$ cs2cs -v
Using from definition: 
Rel. 4.9.3, 15 August 2016
```

Returns:
```
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
1104629.44 7845845.08 0.00
31dE 70dN 0.000
0
```

After these tests it looks like there is not only issues with using etmerc over tmerc from cs2cs, but also using tmerc on cs2cs and tmerc on proj4js. Using tmerc on cs2cs gives the difference of 23 meters, but using tmerc on proj4js gives the difference of 364 meters.

**Comment by perliedman** [comment]
@theashyster I think the 23 meter difference would be ok for you Leaflet use case, and most real world use (since you wouldn't do the project/unproject cycle 10 times), so if we could make Proj4js perform on par with cs2cs, that would likely also be a fix, and possibly an easier one.

**Comment by mbloch** [comment]
Hi @theashyster, you are more than welcome to use `etmerc.js` from `mapshaper-proj` as a starting point, although that won't be much easier than porting directly from the original C. The authors of `proj4js` wanted to use an idiomatic style of JS, while I tried to copy the original C code as much as possible. One difference that you'll see right away is that in `proj4js`, each source code file is a separate Node module, so functions and variables from other files are imported using `require()`.

**Comment by theashyster** [comment]
@perliedman Yes, I think it would be much more acceptable than the difference of 364 meters. Yeah, the project/unproject cycle will probably happen 2 times, so for that the difference is 2.5 meters using tmerc and that is probably quite enough (have not tested in leaflet at the extreme points yet).

Right now I have managed to get proj4js to work on par with mapshaper-proj and using tmerc in both I can achieve the 23 meter difference, but I have managed to do it using basically the code that is used in mapshaper-proj lib for tmerc projection. So I have to work with this a bit, refactor it and try to use proj4js more, but there is progress.

@mbloch Hi, thanks. That could be true, but being able to run the same tests and check the results in almost the same way is definitely easier than doing one thing in js and then comparing that to C. Yeah, a lot of things are written differently and `etmerc.js` looks even more complex than `tmerc.js`.

**Comment by theashyster** [comment]
@perliedman @mbloch Just wanted to let you guys know, that I have managed to use almost the same code that is used in `etmerc.js` in `mapshaper-proj` and it works like a charm. Yesterday I said that `etmerc.js` looks even more complex than `tmerc.js`, but in reality putting it inside of `proj4js` straight from `mapshaper-proj` was really a breeze and much simpler than trying to find the issues with `tmerc.js` in `proj4js`.

I still have some work with this left to polish it a bit and some refactoring to do, but I have already tested it out a bit in `leaflet` and it seems to work great. Probably have to write some tests too, before I can make a pull request with this, so stay tuned.

**Comment by perliedman** [comment]
Keep up the great work, @theashyster !

**Comment by theashyster** [comment]
@perliedman @mbloch @calvinmetcalf Created a pull request which fixes the issues with `tmerc` / `utm` projections https://github.com/proj4js/proj4js/pull/223

**Comment by theashyster** [comment]
@perliedman Hi, just wanted to let you know that the `etmerc` code is merged in the `master` and a new release `2.3.18-alpha` is released which fixes this issue, thanks to @calvinmetcalf 

Doing the same example you provided in the bug report with projecting and unprojecting a coordinate back and forth 10 times now in the UTM projection gives a difference of 9.313225746154785e-10 meters, so almost none, which means that the projection is super stable.

So this can be closed, finally. ✨
